### PR TITLE
feat[contracts]: introduce OVM_GasPriceOracle

### LINKED ISSUE #908: Add congestion price oracle contract
**Is your feature request related to a problem? Please describe.**
The congestion price (used in https://github.com/ethereum-optimism/optimism/issues/823 ) must be propagated to all OE nodes. This will allow replica nodes which sync the sequencer to accurately serve `eth_estimateGas` requests.

**Describe the solution you'd like**
Create a smart contract on L2 which exposes the current congestion price and allows a specified off-chain owner to set it:

- Data Structures:
  - `congestionPrice: uint256`
  - `owner: address`
- Functions:
  - `getCongestionPrice()`
    - RETURNS `congestionPrice`
  - `setCongestionPrice(newCongestionPrice: uint256)`
    - REQUIRES `msg.sender` is equal to `owner`
    - SETS `congestionPrice` to the supplied `newCongestionPrice`

This contract should be included in the predeploys on L2. The owner should be easily configurable.

**Describe alternatives you've considered**
Exposing an endpoint in the sequencer `rollup_getCongestionPrice`. The downside with this approach is it does not scale as cleanly as setting the congestion price in the state because all replicas now need additional information to the rollup chain blocks.

Keeping the congestion price in the state also has the added benefit because in the long term the chain congestion should be tracked in the L2 state & so this is a reasonable first step towards that vision.

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Adds the `OVM_GasPriceOracle` to satisfy #908. Leaving as draft until @karlfloersch confirms this is what he's looking for.

**Metadata**
- Fixes #908

### CODE REVIEW

File: packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_CongestionPriceOracle.sol [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: MIT
```
Comment (tynes) [inline packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_CongestionPriceOracle.sol:1]: This file has a bunch of empty whitespace in it

**Comment by tynes** [comment]
I've opened up a PR on top of this one that includes the fixes that we need to use it in geth, https://github.com/ethereum-optimism/optimism/pull/917

The implementation in geth will have the L2 GasPrice Oracle written in golang that on a timer will read from the state and cache the values in memory

**Comment by snario** [comment]
This needs to re-target `regenesis/0.4.0`.

**Comment by tynes** [comment]
> This needs to re-target `regenesis/0.4.0`.

We will need to do a 1 off deploy of this to mainnet if we want to enable the fee logic without a regenesis. Geth will require a CLI flag which will default to the `0x42..` address that corresponds to this the gas oracle contract but will be able to be set to the contract that we deploy in the short term

**Comment by tynes** [comment]
This needs to have an entry added to https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/src/predeploys.ts

**Comment by tynes** [comment]
This should add a test for the storage slot that the `executionPrice` is held in

**Review by gakonst — COMMENTED** [review]
I've added the missing storage slot test and rebased on regenesis/0.4.0. LGTM otherwise.

**Review by gakonst — APPROVED** [review]

**Pull request merged**